### ISSUE TITLE
Empty List Returned by get_chromecasts

### ISSUE THREAD

**Issue by chrisnorman7** [balloob/pychromecast#117]
Hi,
I've just found pychromecast, and I'm really excited about the idea of using it!!

However when I use get_chromecasts and related functions - including copying the sample code from README.rst - an empty list is returned.

I have tried going back and using discover_chromecasts from discovery.py but with the same effect. I tried with a timeout of 10 seconds to see if bumping up the timeout would fix anything, but it hasn't.

Any idea what is going on?

Thank you very much.

**Comment by sniperlucian** [comment]
Are you using windows?

This might be due to an issue with zeroconf. 

[https://github.com/jstasiak/python-zeroconf/issues/84](url)

**Comment by chrisnorman7** [comment]
Yes it was.


Thank you so much.


Cheers,

**Comment by chibill** [comment]
I am having this same issue but the fixes on the zeroconf do not fix the problem. (I am on windows 10)

**Comment by alxld** [comment]
I'm also having this problem on Windows 10.  I confirmed I can see all my chromecast devices when trying to cast from Chrome on the same machine.  Tried the zeroconf code hack but still I'm getting an empty list returned.

**Comment by lars18th** [comment]
Hi,

This problem is related to my suggestion #149.

I completly hate all "auto-discovering" services that forget to support "manual discovering". You see your Chromecast in your Windows machine because Chrome uses both: mDNS and SSDP discovery.

If the developer improves the search function supporting both protocols, then you can see your devices.
In the meantime, if you need it then you can "manually connect to one Chromecast device": use the function  "pychromecast.Chromecast('chromecast_ip')", so forget to use functions pychromecast.get_chromecasts() & pychromecast.get_chromecast(friendly_name=).

I hope this helps!

**Comment by RcrdBrt** [comment]
If anyone is still affected by this, I ended up with a workaround using [python-nmap](https://bitbucket.org/xael/python-nmap) to discover the ipv4 address by finding the MAC addresses in my local network.

**Comment by lars18th** [comment]
Hi,

I don't know about the last versions. However, is very frustating to rely only on mDNS... it doesn't work in all environment. So, please, provide alternative use of SSDP (the "old" G API for the Chromecast), **and** direct access with the IP address (the autodiscovering is completly useless if you know the IP address of the device).

Regards.

**Comment by RcrdBrt** [comment]
I used the "direct IP" method by discovering it via python-nmap and
creating the object using: `cast = pychromecast.Chromecast(IP_ADDR)`.
The only downside of using nmap with MAC addresses is that you have to run
it with root user.

P. S. : afaik a static IP can't be assigned to a Chromecast.

**Comment by lars18th** [comment]
Hi @RcrdBrt ,

> afaik a static IP can't be assigned to a Chromecast.

You can assign it static using your DHCP server!

> I used the "direct IP" method by discovering it via python-nmap and
> creating the object using: `cast = pychromecast.Chromecast(IP_ADDR)`.

But, for this is **required** that pychromecast has support for "direct IP"... At time, is not possible to use the command line to assign one IP. I'm too fixing the source code to use a fixed IP. However, will be useful to have this option. For example using this idea:

- In the command line you map one IP to some Chromecast name, then in the list of discovered devices the code adds the manual mapping from the command line, and then continue as regular. Using this technique you can add any Chromecast you like, even if is not recognized.

Someone likes to implement it?

**Comment by RcrdBrt** [comment]
Yes @lars18th, sorry I forgot about setting up the DHCP server, I was
thinking client side at that time.

Not sure about your second statement. I'm using the constructor with the IP
of the chromecast and sending videos to it with `play_media()` works
perfectly fine.


On Apr 19, 2017 10:04 AM, "lars18th" <notifications@github.com> wrote:

Hi @RcrdBrt <https://github.com/RcrdBrt> ,

afaik a static IP can't be assigned to a Chromecast.

You can assign it static using your DHCP server!

I used the "direct IP" method by discovering it via python-nmap and
creating the object using: cast = pychromecast.Chromecast(IP_ADDR).

But, for this is *required* that pychromecast has support for "direct
IP"... At time, is not possible to use the command line to assign one IP.
I'm too fixing the source code to use a fixed IP. However, will be useful
to have this option. For example using this idea:

   - In the command line you map one IP to some Chromecast name, then in
   the list of discovered devices the code adds the manual mapping from the
   command line, and then continue as regular. Using this technique you can
   add any Chromecast you like, even if is not recognized.

Someone likes to implement it?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/balloob/pychromecast/issues/117#issuecomment-295151783>,
or mute the thread

**Comment by lars18th** [comment]
Hi @RcrdBrt ,

> Not sure about your second statement. I'm using the constructor with the IP
> of the chromecast and sending videos to it with play_media() works
> perfectly fine.

Yes, sorry! You're right. With the call pychromecast.Chromecast(IP_ADDR), it works.

**Comment by monbru** [comment]
Is possible firewalled mans multicast address  blocked.  Try to open you firewall to receive  224
0.0.1 multicast address.

**Comment by arulrajnet** [comment]
I am using Ubuntu 16.04.4 LTS. pychromecast not able to discover my Android TV even increased `DISCOVER_TIMEOUT = 60` . But it detects direct IP

```
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
>>> import pychromecast
>>> chromecasts = pychromecast.get_chromecasts()
>>> [cc.device.model_name for cc in chromecasts]
[]
>>> cast = pychromecast.Chromecast('192.168.1.143')
>>> print cast
Chromecast('192.168.1.143', port=8009, device=DeviceStatus(friendly_name=u'SHIELD', model_name=u'SHIELD Android TV', manufacturer=u'NVIDIA', uuid=UUID('3ac0312c-5efd-6cee-64ef-6de11a11151a'), cast_type='cast'))
```

**Comment by rkbennett** [comment]
Exact same problem with CentOS and python 3.6

**Comment by elibroftw** [comment]
I encountered this problem by random. It was working fine until it didn't.
Windows 10 Python 3.7.1.

**Comment by elibroftw** [comment]
I made a hacky solution for Windows 10 (subnet_mask = '255.255.255.0')
View it [here](https://gist.github.com/elibroftw/07f9b2b94b099644843489f07b03acde).

**Comment by gallmerci** [comment]
I have a similar issue. I have thread checking regularly available chromecasts discovered by 

`available_casts = pychromecast.get_chromecasts(tries=1, retry_wait=1, timeout=1)`

However, at some point the list is always empty and I can only restart the program to fix it. Any ideas what the problem could be?

**Comment by emontnemery** [comment]
@gallmerci Repeatedly calling `pychromecast.get_chromecasts` will repeatedly open new sockets listening to the mdns port and repeatedly open a connection to each chromecast.

To keep track of present comecasts a `CastListener` should be used, check out the new example here: https://github.com/home-assistant-libs/pychromecast/pull/361

Please also open a new issue about `pychromecast.get_chromecasts` not working when called repeatedly.

**Comment by emontnemery** [comment]
@elibroftw There seems to be problems with mDNS in Windows 10. Maybe this can help you: https://github.com/home-assistant/core/issues/34631#issuecomment-621317649

**Comment by elibroftw** [comment]
@emontnemery Nope, I can't use a solution that requires me to manually modify windows or install additional software since I use `pychromecast` in an application that I ship to other users. 
The solution that worked for me: `zeroconf==0.24.5`

**Comment by emontnemery** [comment]
@elibroftw oh, that must be a bug in zeroconf then. Which versions have you tried except 0.24.5?

**Comment by elibroftw** [comment]
> @elibroftw oh, that must be a bug in zeroconf then. Which versions have you tried except 0.24.5?

The bug had to do with netifaces or something if I recall correctly. I've had this issue for 1 year. I'm tried the versions >=0.25 but with no luck.

**Comment by elibroftw** [comment]
But this only fixes it on my PC. On my laptop, no chromecasts are found.

**Comment by emontnemery** [comment]
@elibroftw do you mean with pychromecast 0.24.5 discovery works on the PC, but it doesnt't work with any zeroconf version on the laptops?

**Comment by elibroftw** [comment]
@emontnemery and it works on my laptop again. Windows issue for sure but not sure what's causing it.
